aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -664,25 +664,20 @@ SparseTensor _coalesce_sparse_cpu(const SparseTensor& self) {
   auto indicesBufferAccessor = indicesBuffer.accessor<int64_t, 1>();
 
   int64_t i = -1;
-  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(at::ScalarType::BFloat16, at::ScalarType::Half, values.scalar_type(),
-                                        "coalesce", [&] {
+  if (values.scalar_type() == at::kBool) {
```

Comment:
Instead of copy pasting the code you could branch on scalar_type before and after `at::native::cpublas::axpy<scalar_t>` within the AT_DISPATCH block using [if_constexpr](https://github.com/pytorch/pytorch/blob/554169fc7b2d9c8c478090612308e85062092b61/c10/util/C%2B%2B17.h#L273-L355).

Reply:
It does not seem to work. I tried inserting
```c++
c10::guts::if_constexpr<std::is_same<scalar_t, bool>::value>([&] { y[i*incy] |= a & x[i*incx]; }, [&] { y[i*incy] += a*x[i*incx]; });
```
into the AT_DISPATCH block but that fails with
```
/home/pearu/git/pytorch/pytorch/aten/src/ATen/native/cpu/BlasKernel.cpp:203:91: error: invalid operands of types 'double' and 'const double' to binary 'operator&'
  203 |         c10::guts::if_constexpr<std::is_same<scalar_t, bool>::value>([&] { y[i*incy] |= a & x[i*incx]; }, [&] { y[i*incy] += a*x[i*incx]; });
      |                                                                                         ~ ^ ~~~~~~~~~
      |                                                                                         |           |
      |                                                                                         double      const double
```
that is, compiler does not remove the if-true block when `scalar_t` is `double`.

Reply:
Now using
```c++
if (type == at::kBool) {
  ...
} else {
  AT_DISPATCH_...
}
```